src/dmd/mars.d

```diff
@@ -769,6 +749,56 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
             }
         }
     }
+
+    // Output the makefile dependencies
+    if (params.emitMakeDeps)
+    {
+        OutBuffer* buf = new OutBuffer();
+
+        // start by resolving and writing the target (which is sometimes resolved during link phase)
+        if (params.link && params.exefile)
+        {
+            escapeMakePath(buf, &params.exefile[0]);
+        }
+        else if (params.lib)
+        {
+            escapeMakePath(buf, library.getFilename());
+        }
+        else if (params.objname)
+        {
+            escapeMakePath(buf, &params.objname[0]);
+        }
+        else if (params.objfiles.length)
+        {
+            escapeMakePath(buf, params.objfiles[0]);
+            foreach (of; params.objfiles[1 .. $])
+            {
+                buf.writestring(" ");
+                escapeMakePath(buf, of);
+            }
+        }
+        else {
+            assert(false, "cannot resolve makedeps target");
+        }
+
+        buf.writestring(":");
+
+        // then output every dependency
+        foreach (dep; params.makeDeps)
+        {
+            buf.writestringln(" \\");
+            buf.writestring("  ");
+            escapeMakePath(buf, dep);
+        }
+        buf.writenl();
+
+        const data = (*buf)[];
+        if (params.makeDepsFile)
+            writeFile(Loc.initial, params.makeDepsFile, data);
+        else
+            printf("%.*s", cast(int)data.length, data.ptr);
+    }
```

Comment:
This giant function is already long enough, so no need to make it longer :P
Joke aside, now that you've consolidated the main output logic to a single place, can you extract this to a separate function?